### PR TITLE
Adding BinCreator, BinCreatorByAbsoluteValue and BinCreatorByStatistics

### PR DESCRIPTION
**Pull request by lakhtanov**

### CODE REVIEW

File: gradient_boosting/binarization/BinCreator.h [hunk 1]
```diff
@@ -1,13 +1,25 @@
 #ifndef GRADIENT_BOOSTING_BINARIZATION_BINCREATOR_H_
 #define GRADIENT_BOOSTING_BINARIZATION_BINCREATOR_H_
 
+#include <vector>
+
 namespace gradient_boosting {
 namespace binarization {
 
+using std::vector;
```
Comment (rialeksandrov) [inline gradient_boosting/binarization/BinCreator.h]: https://google.github.io/styleguide/cppguide.html#Aliases

File: gradient_boosting/binarization/BinCreator.cpp [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#include "gradient_boosting/binarization/BinCreator.h"
+
+namespace gradient_boosting {
+namespace binarization {
+
+vector<double> BinCreator::CreateBins(
+    BinCreator* bin_creator,
```
Comment (rialeksandrov) [inline gradient_boosting/binarization/BinCreator.cpp]: how you are going to create BinCreator object if you don't have ANY public constructors

File: gradient_boosting/binarization/BinCreator.h (hunk 1 above)
Comment (rialeksandrov) [inline gradient_boosting/binarization/BinCreator.h]: https://google.github.io/styleguide/cppguide.html#Aliases

File: gradient_boosting/binarization/BinCreator.h [hunk 3]
```diff
@@ -1,13 +1,25 @@
...
+
 // TODO(lakhtanov): abstract class for creating bins.
 class BinCreator {
  public:
- private:
+  virtual ~BinCreator() = default;
+  static vector<double> CreateBins(
```
Comment (rialeksandrov) [inline gradient_boosting/binarization/BinCreator.h]: how i see this point:
BinCreator which is base creator class should have constructor from vector of borders (it is simplest one - when user knows borders, yep?)
And of course it should have 1 method to bin-ed vector of features.
What should child classes specify? In which way they calc borders

File: gradient_boosting/binarization/BinCreatorByAbsoluteValue.cpp [hunk 4]
```diff
@@ -1,7 +1,37 @@
...
+    return vector<double>();
+  }
+
+  const double min_element =
+      *std::min_element(features.begin(), features.end());
+  const double max_element =
+      *std::max_element(features.begin(), features.end());
+  vector<double> bins(num_bins);
+  if (num_bins == 1) {
+    bins[0] = {(min_element + max_element) / 2};
```
Comment (rialeksandrov) [inline gradient_boosting/binarization/BinCreatorByAbsoluteValue.cpp]: it is really hard to judge implementation when you don't how you bin-ed features, how you use these bins

Comment (rialeksandrov) [reply gradient_boosting/binarization/BinCreatorByAbsoluteValue.cpp]: but it's still interesting and strange!
for {{0,2,4,6,8,10}, 1}  return {5}
for {{0,2,4,6,8,10}, 2}  return {0, 10}
for {{0,2,4,6,8,10}, 3}  return {0, 5, 10}
wat?

File: gradient_boosting/binarization/BinCreator.h (hunk 1 above)
Comment (lakhtanov) [reply gradient_boosting/binarization/BinCreator.h]: Fixed.

File: gradient_boosting/binarization/BinCreatorByAbsoluteValue.cpp (hunk 4 above)
Comment (lakhtanov) [reply gradient_boosting/binarization/BinCreatorByAbsoluteValue.cpp]: Changed logic of this.

File: gradient_boosting/binarization/ThresholdCreator.h [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+class ThresholdCreator {
+ public:
+  explicit ThresholdCreator(size_t num_thresholds);
+  virtual ~ThresholdCreator() = default;
+  static std::vector<double> CreateThresholds(
+      const ThresholdCreator& threshold_creator,
+      const std::vector<double>& features);
+ protected:
+  virtual std::vector<double> CreateThresholds_(
```
[the remark this answers is not in the document]
Comment (lakhtanov) [reply gradient_boosting/binarization/ThresholdCreator.h]: Fixed.

File: gradient_boosting/binarization/ThresholdCreatorByStatistics.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+using std::vector;
+
+vector<double> ThresholdCreatorByStatistics::CreateThresholds_(
+    const vector<double>& features) const {
+  if (features.empty() || !num_thresholds_) {
+    return vector<double>();
+  }
+
+
+  const double delta_statistics = 100.0 / (num_thresholds_ + 1);
```
[the remark this answers is not in the document]
Comment (lakhtanov) [reply gradient_boosting/binarization/ThresholdCreatorByStatistics.cpp]: Fixed.

File: gradient_boosting/binarization/ThresholdCreatorByStatistics.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  std::set<size_t> unique_threshold_feature_positions;
+  for (size_t i = 0; i < num_thresholds_; ++i) {
+    const double threshold_statistics = (i + 1) * delta_statistics;
+    const size_t feature_position =
+        static_cast<size_t>(threshold_statistics / 100 * features.size());
+    unique_threshold_feature_positions.insert(feature_position);
+  }
+
+  vector<double> sorted_features = features;
```
[the remark this answers is not in the document]
Comment (lakhtanov) [reply gradient_boosting/binarization/ThresholdCreatorByStatistics.cpp]: это виртуальный метод, не очень хорошо будет изменять сигнатуру только из-за одного класса, да и разницы нет особой, так что так оставлю

**Pull request merged**